nixos/modules/services/databases/postgresql.nix

```diff
@@ -189,6 +189,19 @@ in
               '';
             };
 
+            passwordFile = mkOption {
+              type = with lib.types; nullOr path;
```

Comment:
I still think that `types.path` for secrets is a huge footgun. As soon as somebody does `${foo}` rather than `toString foo` in a subsequent change somewhere, the file will be written into the store without a warning.

Reply:
Indeed. Would `apply = toString` remedy this?

Reply:
I tested this a little and it does at least cover trivial cases. https://github.com/NixOS/nixpkgs/pull/78640/ would likely be a more complete solution but this works in the meantime.

Reply:
Personally, I don't think it's a good idea to use a syntax construct with the intention of copying sources into the store to specify secrets. At least that's my impression given its behavior.

Does your change solve the core problem I have in mind? I think so, yes.

However, I have the impression that it takes a while until people are aware of the semantics of the path type and until then they easily run into errors like this. It may be implemented correctly here, but there are still enough other wrongly implemented options out there where this is a footgun. And that's understandable given that we work against the semantics of the path-type here IMHO.

The only benefit I see by using the path-type is that we get some kind of validation for free[1] from the Nix parser when using the literal path syntax (the variant of the path-type that I have an issue with to be clear). Personally, I think it's much better to force a string there and then do a `check = x: builtins.substring 0 1 (toString x) == "/"` as it's the case for the string-case in `types.path`.

Is there any other benefit you have in mind?

[1] and normalization of relative paths to absolute ones, but I'd argue that this isn't used that much for cases like this.

Reply:
> I don't think it's a good idea to use a syntax construct with the intention of copying sources into the store to specify secrets.

Could you elaborate on what you mean by this?

There is no intention to copy the secret to the Nix store here whatsoever; exactly the opposite is the case. The `apply` prevents having the specified path being brought into the Nix store even if you declare it as a path literal which would otherwise be a shot in the foot.

There is likely a way you could still break this but it at least covers the trivial case.

> It may be implemented correctly here, but there are still enough other wrongly implemented options out there where this is a footgun.

I'd love a more generic solution but it simply doesn't exist yet and there doesn't appear to have been any progress on that in the past few years.

> The only benefit I see by using the path-type is that we get some kind of validation for free[1] from the Nix parser when using the literal path syntax (the variant of the path-type that I have an issue with to be clear). Personally, I think it's much better to force a string there and then do a `check = x: builtins.substring 0 1 (toString x) == "/"` as it's the case for the string-case in `types.path`.
>
> Is there any other benefit you have in mind?

Nope, merely that the value passed should look like a path. `types.path` is little more than that check anyways and having it be a `types.str` would cause a type error if a path was passed which would be an even better behaviour.

Reply:
Is there a reason there is no `types.secret` that's just an alias to `types.str` with that substring check (and maybe some way to force allow a store path for tests)? It'd be clearer for users and also avoid this problem.

A bit besides the point though, is there an issue this could be spun out into instead of blocking the PR on it? Not that this isn't valuable discussion, but given it's a widespread problem 1) it should be solved generally and 2) the audience here is far smaller than it should be.

Reply:
> Could you elaborate on what you mean by this?

The path literal - that's allowed by types.path - is something being used to copy stuff into the store. At least that's what its semantics are for.

This syntax is allowed by types.path and I think it's wrong to accept this in any case.

> Nope, merely that the value passed should look like a path. types.path is little more than that check anyways and having it be a types.str would cause a type error if a path was passed which would be an even better behaviour.

OK, we're aligned on this I think (given the last change you pushed).

> A bit besides the point though, is there an issue this could be spun out into instead of blocking the PR on it? Not that this isn't valuable discussion, but given it's a widespread problem 1) it should be solved generally and 2) the audience here is far smaller than it should be.

Fair.
Would you be motivated to do so? I may be able to get to it in the upcoming days otherwise.

Reply:
>  Would you be motivated to do so? I may be able to get to it in the upcoming days otherwise.

I'll give it a shot! Should be the kind of change to start with a PR.